executor/mem_reader.go

```diff
@@ -0,0 +1,455 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/distsql"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+)
+
+type memIndexReader struct {
+	baseExecutor
+	index         *model.IndexInfo
+	table         *model.TableInfo
+	kvRanges      []kv.KeyRange
+	desc          bool
+	conditions    []expression.Expression
+	addedRows     [][]types.Datum
+	retFieldTypes []*types.FieldType
+	outputOffset  []int
+	// cache for decode handle.
+	handleBytes   []byte
+	memIdxHandles map[int64]struct{}
```

Comment:
Please add some comments to explain what is this field used for.

Reply:
done.